GoogleForCreators/web-stories-wp
# Performance: SingleSelectionMoveable Improve Render Speed

## Context
Part of looking into `SingleSelectionMoveable` which had the single largest render cost that contributed to ~40ms of the largest commit in updating an elements position.

Recorded interaction was to drag an already selected element to update its position.

| Test Page (50 elements) | Render Cost |
| ---- | ---- |
| <img width="1440" alt="147141603-77df3d34-db93-4ac2-bea7-bcc38644f492" src="https://user-images.githubusercontent.com/35983235/147598529-1eb5916f-7a7b-4512-b959-de8b0c1191a0.png"> | <img width="601" alt="Screen Shot 2021-12-27 at 3 02 53 PM" src="https://user-images.githubusercontent.com/35983235/147597369-bee97d49-624f-483f-9536-0f00e4c9d205.png"> |

Look up findings and summary more in depth here:
https://github.com/google/web-stories-wp/issues/10064#issuecomment-1001765281

## Summary
Fixes render performance of `SingleSelectionMoveable` by updating `useSnapping` & `CanvasProvider`.

| Before | After |
| ---- | ---- |
| ~40ms | ~1ms |
| <img width="599" alt="Screen Shot 2021-12-28 at 12 41 13 PM" src="https://user-images.githubusercontent.com/35983235/147598126-30963119-3cf3-4568-8130-b414b88fe455.png"> |  <img width="599" alt="Screen Shot 2021-12-28 at 12 34 42 PM" src="https://user-images.githubusercontent.com/35983235/147598138-4963dce2-0042-4cfc-b62f-3722c8e9a087.png"> |


## Relevant Technical Choices
Uses a centralized resize observer to update bounding boxes instead of `getBoundingClientRect()` in the hook which attributed to the large style recalc causing `SingleSelectionMoveable` to be such a costly render.

Considered using our `useResizeEffect` but it didn't quite fit for what we needed here:
- useResizeEffect is set up to track a single node, this is extensible to multiple nodes
- useResizeEffect is only used for height and width, this is used for the whole content rect
- useResizeEffect's api is set up to conform to to the effect api, this uses the callback ref pattern @choumx linked in web-stories slack channel

In general, I thought it best to use a custom solution with `ResizeObserver` instead of try to augment our api for `useResizeEffect` to fit this use case. This approach may be useful in other areas, and we can always pull it out and abstract on it, but want to wait till more use cases present themselves instead of pre-emptively doing that

## To-do
If everyone likes the approach with the centralized resize observer, I think there might be a lot of nodes held in state inside `CanvasProvider` that are just held to get their bounding box later. This could replace all those use cases. 

Will create a separate ticket tho if we decide to go with that refactor since it's a separate issue from render perf of SingleSelectionMoveable.

## User-facing changes
NA


## Testing Instructions






This PR can be tested by following these steps:

1. See that snapping while moving an element is not effected by the update


## Reviews

### Does this PR have a security-related impact?



### Does this PR change what data or activity we track or use?



### Does this PR have a legal-related impact?



## Checklist




---



Fixes #10065
